core/pending.go

```diff
@@ -185,3 +217,48 @@ func (p *PreConfirmed) Validate(parent *Header) bool {
 	return p.Block.Number == p.PreLatest.Block.Number+1 &&
 		p.PreLatest.Block.ParentHash.Equal(parent.Hash)
 }
+
+func (p *PreConfirmed) TransactionByHash(hash *felt.Felt) (Transaction, error) {
+	if preLatest := p.GetPreLatest(); preLatest != nil {
+		for _, tx := range preLatest.Block.Transactions {
+			if tx.Hash().Equal(hash) {
+				return tx, nil
+			}
+		}
+	}
+
+	for _, tx := range p.GetCandidateTransaction() {
+		if tx.Hash().Equal(hash) {
+			return tx, nil
+		}
+	}
+
+	for _, tx := range p.GetTransactions() {
+		if tx.Hash().Equal(hash) {
+			return tx, nil
+		}
+	}
+
+	return nil, ErrTransactionNotFound
+}
+
+func (p *PreConfirmed) ReceiptByHash(
+	hash *felt.Felt,
+) (*TransactionReceipt, *felt.Felt, uint64, error) {
+	if preLatest := p.GetPreLatest(); preLatest != nil {
+		for _, receipt := range preLatest.Block.Receipts {
+			if receipt.TransactionHash.Equal(hash) {
+				return receipt, preLatest.Block.Header.ParentHash, preLatest.Block.Number, nil
+			}
+		}
+	}
+
+	preConfirmedBlock := p.GetBlock()
+	for _, receipt := range preConfirmedBlock.Receipts {
```

Comment:
In a similar fashion to Dat's comment above, `p.Block` can be used directly here